presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java

```diff
@@ -422,4 +512,66 @@ private void updatePartitionedStats(
             }
         }
     }
+
+    private Optional<StatisticsFile> getMostRecentStatisticsFileForTableSnapshot(IcebergTableHandle handle)
+    {
+        Snapshot target = handle.getSnapshotId().map(icebergTable::snapshot).orElseGet(icebergTable::currentSnapshot);
+        return icebergTable.statisticsFiles()
+                .stream()
+                .min((first, second) -> {
+                    if (first == second) {
+                        return 0;
+                    }
+                    if (icebergTable.snapshot(first.snapshotId()) == null) {
+                        return 1;
+                    }
+                    if (icebergTable.snapshot(second.snapshotId()) == null) {
+                        return -1;
+                    }
+                    Snapshot firstSnap = icebergTable.snapshot(first.snapshotId());
+                    Snapshot secondSnap = icebergTable.snapshot(second.snapshotId());
+                    long firstDiff = firstSnap.timestampMillis() - target.timestampMillis();
+                    firstDiff = firstDiff < 0 ? Long.MAX_VALUE : firstDiff; // don't files from the past
+                    long secondDiff = secondSnap.timestampMillis() - target.timestampMillis();
+                    secondDiff = secondDiff < 0 ? Long.MAX_VALUE : secondDiff;
+                    return Long.compare(firstDiff, secondDiff);
```

Comment:
Ah good catch. the `*diff` variables should be absolute values

Reply:
I actually think it's better to just use the absolute difference in time, even if a statistic file is in the future. I will update this to use the absolute difference.

Reply:
Great, I agree with you, uniformly consider all snapshots by their absolute distance to the target snapshot would be better.

Is there a possibility to take snapshot's summary message into consideration? For example use "total-records" to calculate absolute row count diff with target, what's your opinion?

Reply:
In [the current Iceberg spec](https://iceberg.apache.org/spec/#snapshots) I don't see a field which has the "total-records" available in the snapshot `summary`. If it exists could you point me to the documentation? I wouldn't be opposed to using something like that but we would need to define a weighting between the time and total records in order to calculate the distance.

My main thought is while record count could be an indication, we would need a reasonable formula to include time-weighting, or else we could make very poor decisions. For example, if there were two snapshots taken a month apart, with many adds/deletes/inserts between the snapshots but they end with the same total records, picking a statistics file based purely on the total records at the time could result in picking a statistics file which is a month old, or a month in the future.

Personally, I think the time weighting here ought to be good enough for most scenarios, but in the case where a user has transactions changing a large number of rows, including a weight considering the total number of records would be beneficial, if it's available.

Reply:
Thanks for the pointer to the code. I didn't realize they had defined properties in the file. If the total record count is available in both snapshots I can implement the record weighting calculation, but there is still the issue of determining a reasonable value for the weighting.

Personally, I don't think I have a good value for this. It would take a fair amount of data and experimentation. If we want to go down this route I would propose that we make it a configurable session property set to a 0-weighting for records by default, such that we only use time in the default case, then a user can configure a weighting if they care more about records or experimentally determine a better value.

The formula for calculating the "total difference" ($\Delta$) between a target snapshot and another snapshot with a configurable weight $\epsilon$, and total records $R$

$$\Delta_\text{snap} = |t_\text{snap} - t_\text{target}| - (\epsilon * |R_\text{snap} - R_\text{target}|)$$

In this case the comparison code would look something like

```java
long recordWeighting = getStatisticRecordCountWeighting(session); // 0 by default

Snapshot firstSnap = icebergTable.snapshot(first.snapshotId());
Snapshot secondSnap = icebergTable.snapshot(second.snapshotId());
long firstDiff = abs(firstSnap.timestampMillis() - target.timestampMillis());
long secondDiff = abs(secondSnap.timestampMillis() - target.timestampMillis());

long snapRecordCount = getRecordCount(target);
long firstRecordDiff = recordWeighting * abs(getRecordCount(firstSnap) - snapRecordCount);
long secondRecordDiff = recordWeighting * abs(getRecordCount(secondSnap) - snapRecordCount);

return Long.compare(firstDiff + firstRecordDiff, secondDiff + secondRecordDiff);
```

Reply:
Sounds great. I agree with you that it makes sense to give users the opportunity to configure the weighting.